### DISCUSSION TITLE
Roster each week for each season

### DISCUSSION THREAD

**Discussion by noahantoun** [cwendt94/espn-api#267] [ESPN API]
Hi is it possible to get each teams roster each week. Even for previous season? I have been playing around and haven't found the ability to do so. Would love to show teams in my league how their team has changed over time. Ideally by player ID.

**Comment by cwendt94** [comment]
You can use the `league.box_scores` which will give you that weeks matchup/roster. You can go through each week and save playerIds from the `home_lineup` and `away_lineup` fields. See [box score class](https://github.com/cwendt94/espn-api/wiki/Box-Score-Class)

**Reply by noahantoun** [reply]
Got it! Thank you! This has been so fun to play around with. Working on a web app for our league. Providing some detailed data for them to look at, but mostly as a fun place to interact with. League/Team overviews, power rankings, historical data, etc. This has been a lot of what ive needed.

**Reply by cwendt94** [reply]
That's a awesome idea, I am glad this package has been helpful!

**Reply by noahantoun** [reply]
Curious if I am approaching this the wrong way. Here is the code I am running to get each teams rosters for the past 3 seasons per week. 
TeamWeekID = []
Week = []
TeamID = []
Roster = []
YearID = []
Year = [2019,2020,2021]

for y in Year:
    from espn_api.football import League
    league = League(league_id= LID, year= y, espn_s2 = ES2')
    for w in range (0,len(league.teams[1].schedule)):
        for m in range(0,len(league.box_scores(week = w))):
            for r in range (0,len(league.box_scores(week = w)[m].home_lineup)):
                YearID.append(y)
                TeamWeekID.append(str(league.box_scores(week = w)[m].home_team) + "-" + str(w))
                Week.append(w+1)
                TeamID.append(league.box_scores(week = w)[m].home_team)
                Roster.append(league.box_scores(week = w)[m].home_lineup[r].playerId)
            for r in range (0,len(league.box_scores(week = w)[m].away_lineup)):
                YearID.append(y)
                TeamWeekID.append(str(league.box_scores(week = w)[m].away_team) + "-" + str(w))
                Week.append(w+1)
                TeamID.append(league.box_scores(week = w)[m].away_team)
                Roster.append(league.box_scores(week = w)[m].away_lineup[r].playerId)
        
rosterweek = {'TeamWeekID':TeamWeekID, 'Year':YearID, 'Week':Week, 'Team ID':TeamID, 'Roster':Roster}
rosterweekdf = pd.DataFrame(teamweek,columns=['TeamWeekID', 'Year', 'Week', 'Team ID', 'Roster'])
rosterweekdf.set_index('TeamWeekID', inplace=True)

Is there a better way to run this? The run time is well over an hour

**Reply by cwendt94** [reply]
Its hard to read without it being formatted but you should only need to call `box_scores` once for a single week. The reason its taking long is because each time you call `box_scores` its calling ESPN's API. Try assigning the box_scores result to a variable then using that for the extra logic.
Something like this
```python3
from espn_api.football import League
for y in Year:
  league = League(.....)
  for w in range(0, len(league.teams[1].schedule)):
    # Set box_score in var so you don't have to call below
    box_score = league.box_scores(week = w)
    for m in range(0, len(box_score)):
      matchup = box_score[m]
      for r in range(0, len(matchup.home_lineup)):
        # Get all of your data here
```
